### DISCUSSION TITLE
setting values on multiple instances of structures ...

### DISCUSSION THREAD

**Discussion by SandyWalsh** [cue-lang/cue#1723] [Q&A]
n00b question. 

Sorry for this somewhat complicated example, but it's a vast simplification of my real-world use-case. Dunno how to boil it down any more. Questions and elaborations in the comments of the snippet. 

```// --- All this stuff would be defined in another package file.

#ProductA : {
    customer: string
    user: string
}

#ProductB : {
    speed: int
    distance: int
}

productA: #ProductA & { customer: #shared.customer, user: #specific.aa}
productB: #ProductB & { speed: #shared.speed, distance: #specific.bb}

// The "items" in Package will be generated by some business process
#Package : {
    #shared: {
        customer: string
        speed: int
    }
    items: [
        {
            #specific : {
                user: "world"
            }
            products: [#ProductA]
        },
        {
            #specific : {
                distance: 999
            }
            products: [#ProductB]
        },
        {
            #specific : {
                user: "world"
                distance: 456
            }
            products: [#ProductA, #ProductB]
        },
    ]    
}

// ---- Now, in another file, I want to make a concrete Foo

order : #Package & {
    #shared: {
        customer: "hello"
        speed: 123
    }
}
```
The errors
```
$ cue export /mnt/cue/sample.cue
thingA.a: reference "#shared" not found:
    ./mnt/cue/sample.cue:13:19
thingA.aa: reference "#specific" not found:
    ./mnt/cue/sample.cue:13:34
thingB.b: reference "#shared" not found:
    ./mnt/cue/sample.cue:14:19
thingB.bb: reference "#specific" not found:
    ./mnt/cue/sample.cue:14:34
*/
```
# The Problem

I want to reuse these "ProductX" structures. They're defined in another file and whomever creates "order" doesn't have know about them. 

The ProductX's should pull from the appropriate context (shared or specific)

Also, since "Package.items" is a list and "order" has no visibility on it (again, in another file) "order" can't be assured of the order of elements so we can't use list indices. 

## Questions

How to make these reusable "Products"? 

How do I pass context into them, like parameters to a function?

**Comment by georgfaust** [comment]
`#shared.customer` does not exist, but `#Package.#shared.customer`
`#specific.aa` does not exist but `#Package.items[0].#specific.user`

You seem to expect changing `#Package` with your `order`, but you do not, look at the output of `cue eval`

Does this help?

```
#ProductA : {
    customer: string
    user: string
}

productA: #ProductA & { 
	customer: #Package.#shared.customer, 
	user: #Package.items[0].#specific.user}

// The "items" in Package will be generated by some business process
#Package : {
    #shared: {
        customer: string
        speed: int
    }
    items: [
        {
            #specific : {
                user: "world"
            }
            products: [#ProductA]
        },
    ]    
}

// ---- Now, in another file, I want to make a concrete Foo

#Package: #shared: customer: "hello"
```

export:

```
{
    "productA": {
        "customer": "hello",
        "user": "world"
    }
}
```

**Reply by SandyWalsh** [reply]
thanks @[georgfaust](https://github.com/georgfaust) let me play with that a little.

**Reply by SandyWalsh** [reply]
so, after playing with this, it's just backwards to the flow I need. I want the products to pull from the context and not push into the products. Using list indicies isn't viable. Difficult to explain. Going back to jsonnet to see if I can get a little further.

**Comment by myitcv** [comment]
I'm afraid I got a bit lost reading through your example 😄 

Perhaps you could try explaining the relationship between things in prose rather than code - that might help the concept to click for me/others?

**Reply by SandyWalsh** [reply]
Sure ... lemme give it a shot. 

We have a bunch of Product definitions that have attributes associated with them. The values for many of these attributes are pulled from values in a Contract. One Contract attribute could be used multiple times across many products. 

For example, a customer signs a contract for new eCar with Red color package. Each aspect of the eCar Product definition needs to pull this "red" value and use it:
* drivers seat color
* passengers seat color
* dashboard & door color
* roof covering color
* rear seats color
* ...

We don't want to have to PUSH the values into each attribute of the Product since the order and quality of attributes will vary greatly from product to product (and some attributes are optional). Instead we want the Product to PULL the values from the Contract. 

The ultimate goal is to be able to look at the Product definition and _infer_ which attributes are missing from the contract. 
For example, they try to sell an eCar with no color package specified, it should complain "ColorPackage value missing" (effectively). If there were 10 attributes missing, a list of the missing attributes. We'd like to use these error messages to inform the UI what to ask for. 

:) did I make it worse or better?

**Reply by verdverm** [reply]
I wonder if the "function" pattern can help your use case?

https://cuetorials.com/patterns/functions/

**Reply by SandyWalsh** [reply]
I don't see the merit of aliasing "in" as "X1" in that use-case? Why not just use "in"? In our use-case, "in" would be "contract"

**Reply by verdverm** [reply]
It may have been related to a CUE bug or something around binding before the data is available. I'll have to look into it, as it seems confusing to me why it is there as well.

You can try ignoring it for now?

The idea was to share the concept as a way to package up the field references, reshaping, and any conditionals you may need.
